metadata/src/main/java/org/apache/kafka/migration/KRaftMigrationDriver.java

```diff
@@ -0,0 +1,414 @@
+                case INACTIVE:
+                    break;
+                case NEW_LEADER:
+                    // This probably means we are retrying
+                    eventQueue.append(new BecomeZkLeaderEvent());
+                    break;
+                case NOT_READY:
+                    break;
+                case ZK_MIGRATION:
+                    eventQueue.append(new MigrateMetadataEvent());
+                    break;
+                case DUAL_WRITE:
+                    eventQueue.append(listener.syncMetadataToZkEvent());
+                    break;
+            }
+
+            // Poll again after some time
+            long deadline = time.nanoseconds() + NANOSECONDS.convert(10, SECONDS);
+            eventQueue.scheduleDeferred(
+                "poll",
+                new EventQueue.DeadlineFunction(deadline),
+                new PollEvent());
+        }
+
+        @Override
+        public void handleException(Throwable e) {
+            log.error("Had an exception in " + this.getClass().getSimpleName(), e);
+        }
+    }
+
+    class MigrateMetadataEvent implements EventQueue.Event {
+        @Override
+        public void run() throws Exception {
+            if (migrationState != MigrationState.ZK_MIGRATION) {
+                log.warn("Skipping ZK migration, already done");
+                return;
+            }
+
+            Set<Integer> brokersWithAssignments = new HashSet<>();
+            log.info("Begin migration from ZK");
+            consumer.beginMigration();
+            try {
+                // TODO use a KIP-868 metadata transaction here
+                List<CompletableFuture<?>> futures = new ArrayList<>();
+                client.readAllMetadata(batch -> futures.add(consumer.acceptBatch(batch)), brokersWithAssignments::add);
+                CompletableFuture.allOf(futures.toArray(new CompletableFuture[]{})).get();
+
+                Set<Integer> brokersWithRegistrations = new HashSet<>(zkBrokerRegistrations.keySet());
+                brokersWithAssignments.removeAll(brokersWithRegistrations);
+                if (!brokersWithAssignments.isEmpty()) {
+                    //throw new IllegalStateException("Cannot migrate data with offline brokers: " + brokersWithAssignments);
+                    log.error("Offline ZK brokers detected: {}", brokersWithAssignments);
+                }
+
+                // Update the migration state
+                OffsetAndEpoch offsetAndEpoch = consumer.completeMigration();
+                apply("Migrating ZK to KRaft", __ -> migrationState(offsetAndEpoch.offset, offsetAndEpoch.epoch));
+            } catch (Throwable t) {
+                log.error("Migration failed", t);
+                consumer.abortMigration();
+            } finally {
+                // TODO Just skip to dual write for now
+                apply("Persist recovery state", client::setMigrationRecoveryState);
+                transitionTo(MigrationState.DUAL_WRITE);
+            }
```

Comment:
What would be idle here? KILL and retry? or simply add MigrateMetadataEvent again technically making async infinite loop in worst case.